Source/WebCore/page/Navigation.cpp

```diff
@@ -135,9 +135,20 @@ void Navigation::initializeForNewWindow(std::optional<NavigationNavigationType>
         if (shouldProcessPreviousNavigationEntries) {
             for (auto& entry : previousNavigation->m_entries)
                 m_entries.append(NavigationHistoryEntry::create(protectedScriptExecutionContext().get(), entry.get()));
-            m_currentEntryIndex = previousNavigation->m_currentEntryIndex;
-            if (navigationType == NavigationNavigationType::Reload) {
-                updateForActivation(currentItem.get(), navigationType);
+
+            // FIXME: This should handle Push, however somewhere before here the currentEntry of previousNavigation was updated
+            // to the new navigation so we get duplicate entries.
+            if (navigationType != NavigationNavigationType::Push) {
+                Ref previousEntry = m_entries[*previousNavigation->m_currentEntryIndex];
+
+                if (navigationType == NavigationNavigationType::Replace)
+                    m_entries[*previousNavigation->m_currentEntryIndex] = NavigationHistoryEntry::create(protectedScriptExecutionContext().get(), *currentItem);
```

Comment:
This seems like it desperately needs a bounds check.  It is not clear to me that this isn't introducing a possible out-of-bounds write.  And a read on line 142.